File: frontend/src/lib/push.js
/**
 * Install and Web Push, in the browser. Spec v2.1 §6 preamble, §4.2, §7.3.
 *
 * §6's preamble is the whole design brief for this file:
 *
 *   "on iPhone, Web Push works only for a PWA added to the home screen (iOS 16.4+), and the
 *    permission request must run inside a user gesture; iOS has no programmatic install
 *    prompt, so member first-run onboarding *guides* Share → Add to Home Screen, detects
 *    standalone mode, and nags until push is granted."
 *
 * So there are two mechanisms, not one with a fallback: Chrome/Edge/Android hand us a real
 * `beforeinstallprompt` event we can spend on a button, and iOS hands us nothing at all. This
 * module reports which world it is in and refuses to pretend — `platform()` returns
 * `ios-safari` where the only honest UI is a set of instructions.
 *
 * Everything here is the browser side of `backend/spielplan/api/push.py`; the endpoint and the
 * auth key go straight from the browser to that route and are never held anywhere else.
 */

import { api, get, post } from '$lib/api.js';

const ua = () => (typeof navigator === 'undefined' ? '' : navigator.userAgent);

/**
 * Which install story applies here.
 *
 *   `ios-safari`   Share → Add to Home Screen, and nothing else exists.
 *   `ios-other`    Chrome/Firefox/Edge on iOS: they cannot add to the home screen at all,
 *                  so the only useful thing to say is "open this in Safari".
 *   `installable`  a `beforeinstallprompt` was captured — a real button.
 *   `browser`      everything else: desktop Safari, Firefox, or a Chrome that has already
 *                  installed the app or does not consider it installable yet.
 *
 * iPadOS 13+ reports itself as a Mac, which is why `maxTouchPoints` is in the test: without
 * it an iPad is told to click an Install button that will never appear.
 */
export function platform({ installPrompt = false } = {}) {
  const agent = ua();
  const ios =
    /iPad|iPhone|iPod/.test(agent) ||
    (typeof navigator !== 'undefined' &&
      navigator.platform === 'MacIntel' &&
      navigator.maxTouchPoints > 1);
  if (ios) {
    // On iOS every browser is WebKit, but only Safari can add a page to the home screen.
    return /CriOS|FxiOS|EdgiOS|OPiOS|GSA/.test(agent) ? 'ios-other' : 'ios-safari';
  }
  return installPrompt ? 'installable' : 'browser';
}

// --- the install prompt, where a browser offers one ------------------------------------------
//
// Chrome and Edge fire `beforeinstallprompt` once per page load and only if they consider the
// app installable; the event is the *only* way to open the real install dialog, and it must be
// kept, because calling `prompt()` outside a user gesture does nothing.
//
// The listener is registered when this module loads rather than when a component mounts, so an
// event that arrives while the member is still on Home is not lost on the way to the account
// screen. It is still best-effort by nature — a browser that fired it before this module was
// ever imported has fired it for good — and the screen renders the honest "no install prompt
// here" branch in that case rather than a button that would do nothing.

/** @type {any} */
let captured = null;
/** @type {Set<(event: any) => void>} */
const watchers = new Set();

if (typeof window !== 'undefined') {
  window.addEventListener('beforeinstallprompt', (event) => {
    // Stops Chrome's own mini-infobar, so the guided act §6 asks for is the only thing on
    // screen instead of racing a browser chrome banner.
    event.preventDefault();
    captured = event;
    for (const watcher of watchers) watcher(event);
  });
}

/** The captured event, or null where this browser has not offered one. */
export const installPrompt = () => captured;

/** @param {(event: any) => void} fn @returns {() => void} unsubscribe */
export function watchInstallPrompt(fn) {
  watchers.add(fn);
  return () => watchers.delete(fn);
}

/**
 * Open the browser's install dialog. Single-use: the event cannot be prompted twice, so it is
 * dropped as it is spent and the button goes away with it.
 */
export async function showInstallPrompt() {
  const event = captured;
  if (!event) return { outcome: 'unavailable' };
  captured = null;
  await event.prompt();
  return (await Promise.resolve(event.userChoice).catch(() => null)) ?? { outcome: 'dismissed' };
}

/** Already installed: launched from the home screen or as a desktop window. */
export function isStandalone() {
  if (typeof window === 'undefined') return false;
  if (window.matchMedia?.('(display-mode: standalone)').matches === true) return true;
  // iOS Safari's own, older flag: it does not implement `display-mode: standalone`, so this is
  // the only signal on the one platform where being installed actually decides whether push
  // works at all. Cast because it is not in lib.dom's `Navigator`.
  return /** @type {any} */ (window.navigator).standalone === true;
}

export const pushSupported = () =>
  typeof window !== 'undefined' &&
  'serviceWorker' in navigator &&
  'PushManager' in window &&
  'Notification' in window;

/**
 * 'default' | 'granted' | 'denied', or 'unsupported' where there is no Notification API.
 *
 * The Permissions API first and `Notification.permission` only as the fallback, because the
 * two do not always agree: a headless Chromium whose notification permission has been granted
 * out of band still reports `Notification.permission === 'denied'`, and a screen that believed
 * that would tell the member their browser is blocking notifications it is perfectly willing
 * to deliver. Safari before 16 has no `notifications` query, which is what the fallback is for.
 */
export async function permissionState() {
  if (!pushSupported()) return 'unsupported';
  try {
    const status = await navigator.permissions?.query({ name: 'notifications' });
    // The Permissions API says 'prompt' where the Notification API says 'default'.
    if (status?.state) return status.state === 'prompt' ? 'default' : status.state;
  } catch {
    // No `notifications` in this browser's permission registry — fall through.
  }
  return Notification.permission;
}

/** What the account screen renders from: the step, the key, and this member's devices. */
export const readState = () => get('/push/state');

export const completeOnboarding = () => post('/setup/onboarding/complete', {});

/**
 * The VAPID application server key, base64url text → the bytes `subscribe()` wants.
 *
 * base64url, not base64: two characters of the alphabet differ and the padding is absent, the
 * same trap `passkeys.js` documents. Getting it wrong produces a subscription the push service
 * will later refuse to deliver to, which is a failure nobody sees for weeks.
 */
export function applicationServerKey(base64url) {
  const b64 = base64url.replace(/-/g, '+').replace(/_/g, '/');
  const raw = atob(b64.padEnd(Math.ceil(b64.length / 4) * 4, '='));
  return Uint8Array.from(raw, (c) => c.charCodeAt(0));
}

/** A name the member will recognise in the device list a year from now. */
export function deviceLabel() {
  const agent = ua();
  if (/iPad/.test(agent)) return 'iPad';
  if (/iPhone/.test(agent)) return 'iPhone';
  if (/Android/.test(agent)) return 'Android phone';
  if (isStandalone()) return 'Installed app';
  return 'This browser';
}

async function currentSubscription() {
  const registration = await navigator.serviceWorker.ready;
  return { registration, subscription: await registration.pushManager.getSubscription() };
}

/**
 * This browser's own endpoint, or null. §6's preamble, §4.2.
 *
 * The one question the account screen could not ask. `GET /api/push/state` answers "which devices
 * does this member have", and `api/push.py`'s own docstring is explicit that it is "this member's
 * devices. Never the household's" — so a screen that derived "notifications are on for THIS
 * device" from that list told a member's second phone that it was already registered, listed the
 * first phone as if it were itself, and hid the only control that can run
 * `Notification.requestPermission()` inside a gesture. Only the browser knows whether the browser
 * holds a subscription, so this is where that fact has to come from. [M4.11 finding 21]
 */
export async function localEndpoint() {
  if (!pushSupported()) return null;
  const { subscription } = await currentSubscription();
  return subscription?.endpoint ?? null;
}

/**
 * Was this subscription minted against the key the server is signing with now?
 *
 * A `PushSubscription` is bound for life to the application server key it was created against, so
 * a pair regenerated on this server — §3.1's first boot after a restore whose `.env` did not come
 * back, or `spielplan-secrets reset` — leaves every phone holding a subscription the push service
 * will refuse to deliver to. Nothing noticed: `enablePush` took whatever `getSubscription()`
 * returned and `syncSubscription` re-posted the same endpoint on every account-page open, so the
 * device list read healthy while no notification could arrive. `send.py` cannot repair it either,
 * deliberately — a key mismatch is a 403 and it does not prune on 403, because a transient
 * rejection must not delete a live phone.
 *
 * **Unknown is treated as a match, never as a mismatch.** `options.applicationServerKey` is absent
 * on older WebKit and on any test double that does not model it, and acting on that absence would
 * unsubscribe a working phone on the strength of a field the browser never implemented. The same
 * goes for a server with no pair at all (`vapidKey` null): there is no key to disagree with.
 */
export function keyMatches(subscription, vapidKey) {
  const held = subscription?.options?.applicationServerKey;
  if (!held || !vapidKey) return true;
  const want = applicationServerKey(vapidKey);
  const have = new Uint8Array(held);
  return have.length === want.length && want.every((byte, i) => byte === have[i]);
}

/**
 * Ask for permission and register this device. **Must be called from a click**: §6's preamble
 * notes the permission request has to run inside a user gesture, and on iOS a request outside
 * one is not deferred — it is refused, permanently, for that origin.
 *
 * Returns `{ permission, subscribed }` rather than throwing on a refusal: "no" is an answer,
 * not an error, and §3.1's fifth step completes either way. On a success it also returns this
 * device's own `endpoint` and the server's `device` handle for it, so the caller can say which row
 * in the member's list is the phone in the member's hand without hashing anything: §2 puts the app
 * behind one plain-HTTP port, where `crypto.subtle` does not exist.
 *
 * @param {{vapidKey?: string|null}} [options]
 */
export async function enablePush({ vapidKey = null } = {}) {
  if (!pushSupported()) return { permission: 'unsupported', subscribed: false };

  const answer = await Notification.requestPermission();
  if (answer !== 'granted') return { permission: answer, subscribed: false };

  const { registration, subscription } = await currentSubscription();
  let held = subscription;
  if (held && !keyMatches(held, vapidKey)) {
    // A subscription bound to a retired key cannot be repaired, only replaced — `subscribe()`
    // with a different `applicationServerKey` is an InvalidStateError while the old one is live,
    // so the old one goes first. This is the act the member just asked for, so it is also the
    // only moment at which silently dropping a subscription is honest. [M4.11 finding 21]
    await held.unsubscribe();
    held = null;
  }
  const live =
    held ??
    (await registration.pushManager.subscribe({
      // Required by every browser that implements Push: a push may not be silent. It is also
      // the honest description of what this is for — §7.3's prompt is a visible question.
      userVisibleOnly: true,
      ...(vapidKey ? { applicationServerKey: applicationServerKey(vapidKey) } : {})
    }));

  const state = await post('/push/subscribe', {
    ...live.toJSON(),
    device_label: deviceLabel()
  });
  return {
    permission: 'granted',
    subscribed: true,
    endpoint: live.endpoint,
    device: state.device,
    subscriptions: state.subscriptions
  };
}

/**
 * Re-post the subscription this browser already holds.
 *
 * A phone re-registers its service worker on every app update and the push service may hand
 * back the same endpoint or a fresh one; either way the server row must match what the browser
 * actually has. The route upserts on the endpoint (§4.2's UNIQUE), so calling this on every
 * open is cheap and cannot fan out into duplicate rows — which is exactly why it upserts.
 *
 * It no longer re-posts a subscription minted under a key this server has replaced. That re-post
 * was the thing that made the fault invisible: the row kept coming back fresh on every account-page
 * open, so the screen said "on" for a device that could not be reached. `{ stale: true }` is the
 * answer instead, and the screen reads it as off-with-a-reason — the member's own tap is what
 * replaces the subscription (`enablePush`), because only a tap is the gesture §6's preamble
 * requires. [M4.11 finding 21]
 *
 * @param {{vapidKey?: string|null}} [options]
 */
export async function syncSubscription({ vapidKey = null } = {}) {
  if ((await permissionState()) !== 'granted') return null;
  const { subscription } = await currentSubscription();
  if (!subscription) return null;
  if (!keyMatches(subscription, vapidKey)) return { stale: true, subscriptions: null };
  return post('/push/subscribe', { ...subscription.toJSON(), device_label: deviceLabel() });
}

/**
 * Turn notifications off for this device: the server row first, then the browser's own
 * subscription. In that order — a browser subscription dropped while the row survives is a
 * target the sender will keep writing to and the push service will keep rejecting.
 *
 * `{ removed: false, subscriptions: null }` rather than a bare `null` when this browser holds
 * nothing to delete. The bare null was indistinguishable from a successful delete that returned no
 * list, and the caller resolved the ambiguity the worst way available — `state?.subscriptions ?? []`
 * blanked the device list, so the screen claimed notifications were off for the household while the
 * member's other phone stayed subscribed and kept receiving them. Two facts, said separately:
 * whether a row was deleted, and what the member's remaining devices are. [M4.11 finding 21]
 */
export async function disablePush() {
  if (!pushSupported()) return { removed: false, subscriptions: null };
  const { subscription } = await currentSubscription();
  if (!subscription) return { removed: false, subscriptions: null };
  const state = await api('/push/subscription', {
    method: 'DELETE',
    body: { endpoint: subscription.endpoint }
  });
  await subscription.unsubscribe();
  return { removed: true, subscriptions: state?.subscriptions ?? null };
}

File: backend/spielplan/api/push.py
"""Web-push subscriptions. Spec v2.1 §4.2, §6 preamble, §7.3, §12 (M2).

§4.2 carries the table and the worker already prunes it; what M2 owes is the write that fills
it. §7.3's finish prompt "queues and surfaces as an in-app banner" *when undeliverable* — a
fallback that only means something once the deliverable path exists.

Three properties this module is built around, all of them from §6's preamble and §4.2:

  * **One row per endpoint.** A phone re-registers its service worker on every app update and
    resubscribes with the same endpoint each time; if that inserted, a household of two phones
    would accumulate a row per release and every notification would arrive n times. `endpoint`
    is UNIQUE in the schema, so the upsert is on the endpoint, not on (user, endpoint).
  * **A subscription belongs to the member, not to the household.** §4.2 keys the table on
    `user_id`, and §7.3's prompt is "per-user": every read here is scoped to the signed-in
    member, so one member's phone can never be handed another member's notification.
  * **The keys are secrets.** The endpoint URL is a bearer capability — anyone holding it can
    push to that device — and `auth` is the message-encryption key. Neither is logged, and
    neither comes back out of the API. Devices are identified to the UI by a hash of the
    endpoint instead.
  * **The two writes need the account credential, not the PIN** (decision 170). §3.2 makes the
    PIN a convenience for the household phone, so a PIN session reaches the product surfaces and
    nothing that changes how the account is reached later. Subscribing rebinds an endpoint to the
    caller and unsubscribing silences a device: whoever holds a switched-into phone could point
    another member's notifications at it, which is the same class as minting a passkey. Reading
    this member's own devices stays `ActiveUser` — it discloses nothing the phone does not hold.

And one property that is not about the table at all: **a stored endpoint is a URL this server
will POST to.** `push/send.py` sends to it verbatim, so what `POST /api/push/subscribe` accepts
is what the backend can be made to reach (§14.3, sec-13). The scheme and address rules are on
`SubscriptionIn`; the one that needs a connection — the household's own servers — is in the
route.

The *sending* half is M4's `spielplan.push` (§12); this is the subscribe/unsubscribe path plus
the read the onboarding screen needs — and `device_handle` now comes from `push/send.py`, where
a rule about naming a device belongs (CLAUDE.md, arch-02). `router` is exported for
`spielplan.app` to register.
"""

from __future__ import annotations

import ipaddress
import logging
from urllib.parse import urlsplit

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel, Field, field_validator

from spielplan.api.deps import DB, ActiveUser, CredentialedUser
from spielplan.connectors.registry import JELLYFIN
from spielplan.core.config import settings
from spielplan.push import keys
from spielplan.push.send import device_handle

router = APIRouter(prefix="/api/push", tags=["push"])
log = logging.getLogger(__name__)


class SubscriptionKeys(BaseModel):
    """The two keys `PushSubscription.toJSON()` carries, under the names the browser uses."""

    p256dh: str = Field(min_length=1, max_length=256)
    auth: str = Field(min_length=1, max_length=256)


class SubscriptionIn(BaseModel):
    # Shaped so the client can post `subscription.toJSON()` plus a label, unchanged. Extra
    # fields (`expirationTime`) are ignored rather than rejected: browsers add to that object.
    endpoint: str = Field(min_length=1, max_length=2048)
    keys: SubscriptionKeys
    device_label: str | None = Field(default=None, max_length=64)

    @field_validator("endpoint")
    @classmethod
    def _endpoint_is_a_push_service_on_the_internet(cls, value: str) -> str:
        """§14.3: a stored endpoint is a URL this server later POSTs to, signed (sec-13).

        `push/send.py` posts to the stored string verbatim with the VAPID JWT attached, so until
        this check existed `POST /api/push/subscribe` was a way for any member's session to make
        the backend knock on an address of the caller's choosing inside the household's network —
        a reachability probe, since the method is fixed and the body is ciphertext, but §14.3
        does not grade the ones that only probe. A length bound was the whole contract.

        The two hosts in play are easy to confuse and only one of them is constrained here. §2
        puts the *app* behind "one plain-HTTP port", which is why `send.py::vapid_subject` exists
        at all: `PUBLIC_URL` is `http://...` on a LAN or Tailscale install and RFC 8292 will not
        take that as a `sub`. This rule is about the *push service's* endpoint instead — minted
        by Apple, Google or Mozilla, always https, never on the household's own network — so
        requiring https here says nothing about how the household reaches the app.

        A literal address is judged by `ipaddress`; a name is not resolved. Resolution happens at
        send time in a different process and a different second, so a validator that pre-empted
        it would be both wrong and a second lookup. No allowlist of push-service domains: the
        set of services is the set of browsers, it changes without asking this household, and an
        allowlist that lags one release silently stops notifications (the failure §4.2's pruning
        rules are written to avoid).

        The literal is judged by REACHABILITY (`is_global`) and not by an enumeration of ranges.
        An enumeration is a list that has to be maintained against `ipaddress` itself: written as
        `is_loopback or is_private or is_link_local or is_reserved or is_unspecified` it let
        100.64.0.0/10 through, because CVE-2024-4032 took RFC 6598 shared address space out of
        `is_private` in 3.12.4 — and that block is exactly what Tailscale assigns every node on a
        tailnet, which is the deployment the paragraph above names. `is_global` is one predicate
        that covers all five and that block, and it says the thing the rule means: a push service
        is on the internet. [M4.11 review cycle 2: m411-c2-push-02; §14.3]
        """
        parts = urlsplit(value)
        if parts.scheme != "https":
            raise ValueError("a push endpoint must be an https URL")
        host = (parts.hostname or "").strip().rstrip(".")
        if not host:
            raise ValueError("a push endpoint must name a host")
        if host.lower() == "localhost":
            raise ValueError("a push endpoint may not name this machine")
        try:
            address = ipaddress.ip_address(host)
        except ValueError:
            return value
        if not address.is_global:
            raise ValueError("a push endpoint may not name a private or local address")
        return value


class EndpointIn(BaseModel):
    # Deliberately not validated the way `SubscriptionIn.endpoint` is: this is the body of a
    # DELETE, and a row stored before that rule existed must stay deletable by the string it
    # was stored under. Refusing to *forget* an endpoint protects nothing.
    endpoint: str = Field(min_length=1, max_length=2048)


async def _refuse_a_household_host(conn, endpoint: str) -> None:
    """The half of sec-13 that needs a database read, which is why it is not in the validator.

    Two hosts this household runs are perfectly public names and would therefore survive
    `SubscriptionIn`'s check: the Jellyfin server (§3.3's connector, reachable from this process
    with an admin key) and the app itself (`PUBLIC_URL`). Neither is ever a push service, so an
    endpoint naming one is a probe by construction rather than a configuration mistake.

    The URL is read straight out of `connector_config` rather than through
    `registry.load_jellyfin`, because that opens the connector's secrets: §3.1 makes a boot with
    an unreadable DEK legal and subscribing a phone must not be the route that fails on it.
    """
    host = (urlsplit(endpoint).hostname or "").strip().rstrip(".").lower()
    configured = await conn.fetchval(
        "SELECT config ->> 'url' FROM connector_config WHERE name = $1", JELLYFIN
    )
    ours = {
        (urlsplit(str(configured or "")).hostname or "").strip().rstrip(".").lower(),
        settings().rp_id.strip().rstrip(".").lower(),
    }
    if host and host in ours:
        raise HTTPException(
            status.HTTP_422_UNPROCESSABLE_ENTITY,
            "a push endpoint may not name this household's own servers",
        )


async def vapid_public_key(conn) -> str | None:
    """The application server key the browser needs at subscribe time, or None.

    Chrome and Edge refuse `pushManager.subscribe()` without one. M2 read `VAPID_PUBLIC_KEY`
    from the environment because the sender that owns the pair was still M4 work; that env var
    is gone with this milestone, and deliberately not kept as an override. A subscription is
    bound to the key it was created against, so an env-supplied public half whose private half
    nothing holds would let every browser subscribe against a key `push/send.py` cannot sign
    with — the push service would reject every delivery and no screen would ever say so. §2
    puts the pair in the database ("generated at first boot and stored the same way"), and the
    database is now the only answer.

    Still None when the pair is absent (§3.1's half-configured boot), which the onboarding
    screen renders as "notifications aren't configured yet" rather than throwing a
    DOMException at the member.
    """
    return await keys.public_key(conn)


async def _subscriptions(conn, user_id: int) -> list[dict[str, object]]:
    """This member's devices. Never the household's — see the module docstring."""
    rows = await conn.fetch(
        """
        SELECT id, device_label, endpoint, created_at, last_seen_ok
        FROM push_subscription WHERE user_id = $1 ORDER BY created_at
        """,
        user_id,
    )
    return [
        {
            "id": row["id"],
            "device_label": row["device_label"],
            # Deliberately not the endpoint: see `device_handle`.
            "device": device_handle(row["endpoint"]),
            "created_at": row["created_at"],
            "last_seen_ok": row["last_seen_ok"],
        }
        for row in rows
    ]


async def _onboarding_done(conn, user_id: int) -> bool:
    """§3.1's fifth step, read back per user.

    `POST /api/setup/onboarding/complete` records completion as `{user_id: true}` inside one
    `setup_step` row, because §6's preamble makes onboarding a per-phone act — one member
    finishing it must not silence the prompt for the other. Nothing read that back until now,
    which is why the screen could not know whether to ask.
    """
    detail = await conn.fetchval("SELECT detail FROM setup_step WHERE step = 'onboarding'")
    return bool((detail or {}).get(str(user_id)))


@router.get("/state")
async def state(user: ActiveUser, conn: DB) -> dict[str, object]:
    """Everything the onboarding screen needs, in one round trip."""
    return {
        "onboarding_complete": await _onboarding_done(conn, user.id),
        "vapid_public_key": await vapid_public_key(conn),
        "subscriptions": await _subscriptions(conn, user.id),
    }


@router.post("/subscribe", status_code=status.HTTP_201_CREATED)
async def subscribe(body: SubscriptionIn, user: CredentialedUser, conn: DB) -> dict[str, object]:
    """Store a browser PushSubscription for the signed-in member.

    ON CONFLICT on `endpoint` rather than an insert: a phone resubscribes with the same
    endpoint every time its service worker is replaced, and §4.2's UNIQUE says so — one
    endpoint is one device, and one device is one row.

    The conflicting row's `user_id` is overwritten on purpose. An endpoint is minted per
    browser profile, so the only way an existing endpoint arrives under a different member is
    that the same browser is now signed in as that member — and the notifications must follow
    the person at the phone, not the person who first granted permission on it.

    `last_seen_ok` is left ALONE on the conflict path, and that is what keeps §4.2's 90-day prune
    off a live phone. The column means "the push service accepted a delivery", so a genuinely new
    endpoint inserts with it NULL and is judged by `created_at = now()` — which is the row the
    prune was written for. Resetting it on a re-post looked like the conservative reading and was
    the opposite once `worker.py` started pruning on `COALESCE(last_seen_ok, created_at)`:
    `push.js::syncSubscription` re-posts the unchanged
    endpoint on every /account open, and on a household older than ninety days that made opening
    the account page arm the deletion of a live device on the same night's prune — measured end
    to end. Keeping the real delivery time is what makes the COALESCE harmless.
    [review cycle 1: m411-rev1-push-prune-deletes-a-live-device-that-just-re-registered; §4.2]

    `created_at` IS rewritten on the conflict path, and that is the other half of the same
    composition. The prune's own docstring says the row it takes is one "whose device never came
    back at all - no delivery, no rejection, ninety days of silence", and a re-post is the device
    coming back: `pushManager.getSubscription()` hands the browser's endpoint back only while that
    subscription is live, so an endpoint that CONFLICTS is a live device by construction — a lapsed
    one mints a new endpoint and INSERTs. With neither column moving, a household that simply sends
    no pushes for ninety days — no Jellyfin link arms no §7.3 prompt, and §6.2's invitation never
    goes to the host — lost every subscription on schedule however often the member re-registered,
    and `send_to_user` then returned `[]` in silence. The cost is that §6.6's device list, ordered
    by `created_at`, now reads in order of last registration rather than first, which is the more
    useful order for the thing that list is for. [M4.11 review cycle 2: m411-c2-push-01; §4.2]

    `SubscriptionIn` has already refused anything that is not an https host off this network;
    what is left is the half of that rule which needs a connection (sec-13).
    """
    await _refuse_a_household_host(conn, body.endpoint)
    handle = device_handle(body.endpoint)
    row = await conn.fetchrow(
        """
        INSERT INTO push_subscription (user_id, device_label, endpoint, p256dh, auth)
        VALUES ($1, $2, $3, $4, $5)
        ON CONFLICT (endpoint) DO UPDATE SET
            user_id      = EXCLUDED.user_id,
            device_label = COALESCE(EXCLUDED.device_label, push_subscription.device_label),
            p256dh       = EXCLUDED.p256dh,
            auth         = EXCLUDED.auth,
            created_at   = now()
        RETURNING id, created_at
        """,
        user.id,
        body.device_label,
        body.endpoint,
        body.keys.p256dh,
        body.keys.auth,
    )
    # The handle, never the endpoint or the auth key.
    log.info("push subscription stored for user %s (device %s)", user.id, handle)
    return {
        "ok": True,
        "id": row["id"],
        "device": handle,
        "subscriptions": await _subscriptions(conn, user.id),
    }


@router.delete("/subscription")
async def unsubscribe(body: EndpointIn, user: CredentialedUser, conn: DB) -> dict[str, object]:
    """Drop one device, scoped to the member who owns it.

    The `user_id = $1` in the WHERE clause is the load-bearing half: without it, an endpoint
    string is enough to un-subscribe somebody else's phone, and endpoints travel (a shared
    browser, a copied bug report). A row belonging to another member is not "forbidden" here
    but simply not found, which is the same answer as a row that never existed.
    """
    removed = await conn.fetchval(
        "DELETE FROM push_subscription WHERE user_id = $1 AND endpoint = $2 RETURNING id",
        user.id,
        body.endpoint,
    )
    if removed is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "no such device on this account")
    log.info("push subscription removed for user %s (device %s)", user.id, device_handle(body.endpoint))
    return {"ok": True, "subscriptions": await _subscriptions(conn, user.id)}
